### ISSUE TITLE
ContextualToolbar should be not displayed when all components inside are disabled

### ISSUE THREAD

**Issue by oskarwrobel** [ckeditor/ckeditor5-ui#269]
This will automatically turn off CT when the editor will be in read-only mode (see read-only requirements https://github.com/ckeditor/ckeditor5/issues/492). And this might be a nice way to disable toolbar for specific elements.

**Comment by oskarwrobel** [comment]
Or maybe is too generic and toolbar should be disabled only when the editor is in read-only mode? @oleq WDYT?

**Comment by oleq** [comment]
I don't think it should be disabled when in read–only as it may contain some features which are always relevant no matter the state (e.g. fullscreen, clipboard, custom user buttons, etc.). We may accidentally block features important for the users by doing this that way. I'm strongly against this approach.

The first idea, which would disable the CT when all buttons are disabled seems a lot safer. The only issue I see here is discoverability and general UX of this approach. With the toolbar gone, it's hard to tell whether the editor is either:
* broken,
* in read–only mode,
* working but features that have their representation in the toolbar are disabled.

Basically, there's no feedback telling people what is going on. It may be very confusing and frustrating because users may want to type but they can't and they don't know why. 

The toolbar, even if disabled, explains a lot. E.g. if they wanted to link and they couldn't because keystroke doesn't work, they will see that the button is disabled and everything becomes clear. Without the toolbar that announces the state of the features things become unclear.

**Comment by pjasiun** [comment]
> I don't think it should be disabled when in read–only as it may contain some features which are always relevant no matter the state (e.g. fullscreen, clipboard, custom user buttons, etc.). We may accidentally block features important for the users by doing this that way. I'm strongly against this approach.

I agree that we should check if all buttons are disabled before we decided not to show the toolbar. The problem is that at this point there is nothing like "toolbar item interface", so not all items will have `disabled` property. I think that we could introduce some interface.

> The first idea, which would disable the CT when all buttons are disabled seems a lot safer. The only issue I see here is discoverability and general UX of this approach. With the toolbar gone, it's hard to tell whether the editor is either:
> 
> broken,
> in read–only mode,
> working but features that have their representation in the toolbar are disabled.
> Basically, there's no feedback telling people what is going on. It may be very confusing and frustrating because users may want to type but they can't and they don't know why.

I think that it's integration what should ensure the proper UX. The one who set the editor to be in the read-only mode should show the user what is going on: create some kind of notification. However, since it might be not only editor what is not editable, the notification (UX in general) could be outside the editor.

> The toolbar, even if disabled, explains a lot. E.g. if they wanted to link and they couldn't because keystroke doesn't work, they will see that the button is disabled and everything becomes clear. Without the toolbar that announces the state of the features things become unclear.

I don't think so. If there is no proper information provided by the integration it is not clear what is going on in both ways: I don't now why I can't link and I don't know why buttons are disabled.

**Comment by pjasiun** [comment]
I tested how the contextual toolbar looks like in the read-only mode with all buttons disabled and it does not look good. My feeling was something in between: "Why I see this toolbar if I can do nothing with it?" and "The system disabled buttons but there is nothing more done than necessary". Especially since it was in contrast to the other part of the application which changed its interface.

**Comment by oleq** [comment]
@fredck, can you [chime in](https://github.com/ckeditor/ckeditor5-image/issues/123#issuecomment-313639614)?

> I'm for not showing contextual toolbars at all, if the editor is readonly.

I'd like to know your opinion about my arguments https://github.com/ckeditor/ckeditor5-ui/issues/269#issuecomment-313646405.

Still, my opinion is that if we put the balloon toolbar editor in read–only mode and disable the toolbar (even if all buttons are disabled) it will become a wall of text without any kind of UI, which may look like the editor is broken. Even if the integration applies some gray background to editable in such situation. 

The read–only mode should have an impact on the state of the UI, not on whether it is displayed or not.

> I don't think so. If there is no proper information provided by the integration it is not clear what is going on in both ways: I don't now why I can't link and I don't know why buttons are disabled.

@pjasiun But you don't spend your time making the toolbar show up to no avail and wondering what's going on because it stopped working.

----

Anyway, **if** we decide to hide the toolbar in read–only:

> I agree that we should check if all buttons are disabled before we decided not to show the toolbar. The problem is that at this point there is nothing like "toolbar item interface", so not all items will have disabled property. I think that we could introduce some interface.

That's true. We've never established such interface. OTOH we silently agreed that toolbar items (well...  so far we have only the dropdown and the button) have `isEnabled` attribute, which controls the thing. So we may assume that it's a good start. 

What we need is `ToolbarView#getItems()`, which returns an array:

```
const toolbar = editor.ui.view.top.get( 0 );

if( !toolbar.getItems().some( item => item.isEnabled ) ) {
    // let's hide the toolbar, there's nothing enabled in there
}
```

**Comment by fredck** [comment]
> @fredck, can you chime in?

I still think the toolbar should not be shown.

What we have there is a contextual toolbar, which should show actions that make sense for the current context. In a read-only situation, the toolbar is totally pointless.

I also understand the fact that CKEditor 5 itself doesn't have any UI that can help the user understand that the editor is in read-only. This feature should be left to the application though, because that's the one that decided to put the editor in such mode, for its own reasons. 

For example, one application may really want the editor to be read-only, so the user will just read the content. Here the toolbar would be totally senseless. Another may want to do so to react to a missing connection to the Internet. In such case, the application itself can show to user an alert about such missing connection.

And in all cases, the disabled buttons in the toolbar can also give the feeling that something is broken in the editor, so it doesn't solve things much.
